re-use HostDirFromRoots for both local pulls and transfer service pulls

rpardini opened the pull request:
- move (and export) `HostDirFromRoots` to core/remotes/docker/config (together with `HostDirFromRoot`)
- in transfer/registry, use `filepath.SplitList()` and use shared `HostDirFromRoots`
- this fixes `config_path = '/etc/containerd/certs.d:/etc/docker/certs.d'` (the default) when used with the transfer service (also default)
- ignore empty roots so leading/trailing/repeated path-list separators don't probe relative CWD paths; return `ErrNotFound` when no valid roots remain
- document `WithHostDir` path-list support (os.PathListSeparator)
- add `TestHostDirFromRoots` covering ordering, precedence, not-found, and empty/nil roots

rpardini commented:
Realized way too late:
- Fixes #12415 
- extremely similar to #12481

samuelkarp commented on core/remotes/docker/config/hosts.go:
```diff
@@ -287,6 +287,22 @@ func updateTLSConfigFromHost(tlsConfig *tls.Config, host *hostConfig) error {
+	for i := range roots {
+		rootfn[i] = HostDirFromRoot(roots[i])
+	}
+	return func(host string) (dir string, err error) {
```
This seems worth fixing.

rpardini replied:
Indeed. Fixed. Thanks.

rpardini committed:
re-use HostDirFromRoots for both local pulls and transfer service pulls

- move (and export) `HostDirFromRoots` to core/remotes/docker/config (together with `HostDirFromRoot`)
- in transfer/registry, use `filepath.SplitList()` and use shared `HostDirFromRoots`
- this fixes `config_path = '/etc/containerd/certs.d:/etc/docker/certs.d'` (the default) when used with the transfer service (also default)
- ignore empty roots so leading/trailing/repeated path-list separators don't probe relative CWD paths; return ErrNotFound when no valid roots remain
- document WithHostDir path-list support (os.PathListSeparator)
- add TestHostDirFromRoots covering ordering, precedence, not-found, and empty/nil roots

Signed-off-by: Ricardo Pardini <ricardo@pardini.net>

rpardini commented on core/remotes/docker/config/hosts.go:
```diff
@@ -287,6 +287,22 @@ func updateTLSConfigFromHost(tlsConfig *tls.Config, host *hostConfig) error {
 	return nil
 }
 
```
Done, thanks.

rpardini commented on core/transfer/registry/registry.go:
```diff
@@ -125,7 +126,7 @@ func NewOCIRegistry(ctx context.Context, ref string, opts ...Opt) (*OCIRegistry,
 
 	hostOptions := config.HostOptions{}
 	if ropts.hostDir != "" {
-		hostOptions.HostDir = config.HostDirFromRoot(ropts.hostDir)
+		hostOptions.HostDir = config.HostDirFromRoots(filepath.SplitList(ropts.hostDir))
 	}
```
Done. Thanks.

rpardini commented on core/transfer/registry/registry.go:
```diff
@@ -125,7 +126,7 @@ func NewOCIRegistry(ctx context.Context, ref string, opts ...Opt) (*OCIRegistry,
 
 	hostOptions := config.HostOptions{}
 	if ropts.hostDir != "" {
-		hostOptions.HostDir = config.HostDirFromRoot(ropts.hostDir)
+		hostOptions.HostDir = config.HostDirFromRoots(filepath.SplitList(ropts.hostDir))
 	}
```
Indeed. Thanks.

rpardini commented on core/remotes/docker/config/hosts.go:
```diff
@@ -287,6 +287,22 @@ func updateTLSConfigFromHost(tlsConfig *tls.Config, host *hostConfig) error {
 	return nil
 }
 
+func HostDirFromRoots(roots []string) func(string) (string, error) {
+	rootfn := make([]func(string) (string, error), len(roots))
+	for i := range roots {
+		rootfn[i] = HostDirFromRoot(roots[i])
+	}
+	return func(host string) (dir string, err error) {
+		for _, fn := range rootfn {
+			dir, err = fn(host)
+			if (err != nil && !errdefs.IsNotFound(err)) || (dir != "") {
+				break
+			}
+		}
+		return
+	}
+}
```
Added tests. Thanks.

rpardini commented:
Addressed the reviews, rebased onto main; CoPilot is now quiet. A Fuzzing test failure, but I don't think it is related to this change.

thaJeztah commented on core/transfer/registry/registry.go:
```diff
@@ -73,7 +74,10 @@ func WithCredentials(creds CredentialHelper) Opt {
 }
 
-// WithHostDir specifies the host configuration directory.
+// WithHostDir specifies the host configuration directory. The value may be a
+// single directory or a list of directories separated by os.PathListSeparator
+// (e.g. "/etc/containerd/certs.d:/etc/docker/certs.d" on Unix), which are
+// searched in order for a matching host configuration.
 func WithHostDir(hostDir string) Opt {
```
Not sure if I like changing this, as it changes semantics of this function a lot, and I'm not sure if we should make platform (unix/windows) PATH semantics make their way through here.

Looking at the change that added support for "multiple paths"; https://github.com/containerd/containerd/commit/b0b6d9aa03ccf903f9bf1a9bc80ea2eecb0b22e9


I (personally) think that may have been a bad choice; it codifies platform-specific formats in the config-file. On Linux, `:` is used as path separator, which cannot be used on Windows (as `C:\....`), and on Windows `;` is used.

Perhaps the preferred option should be to deprecate `Registry.ConfigPath`, and make the configuration itself accept an ordered list of paths to consider.

rpardini replied:
While I agree, this PR is _not_ changing the semantics -- they come from `filepath.SplitList` introduced in 2022. This is just fixing a real, palpable bug that comes from the now-default transfer service _not_ respecting the previous semantics (and not-splitting via the separator): #12415 (and impacting many projects, eg https://github.com/spegel-org/spegel/issues/1090)

Let me know, as this is six-months-old by now, and implementing such a change is way beyond the scope here.
